internal/self_metrics/self_metrics.go

```diff
@@ -0,0 +1,140 @@
+// Copyright 2022 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package self_metrics
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	mexporter "github.com/GoogleCloudPlatform/opentelemetry-operations-go/exporter/metric"
+	meter "go.opentelemetry.io/otel/metric"
+	"github.com/GoogleCloudPlatform/ops-agent/confgenerator"
+	"go.opentelemetry.io/contrib/detectors/gcp"
+	"go.opentelemetry.io/otel/attribute"
+	"go.opentelemetry.io/otel/metric/instrument"
+	"go.opentelemetry.io/otel/sdk/metric/controller/basic"
+	"go.opentelemetry.io/otel/sdk/metric/sdkapi"
+	"go.opentelemetry.io/otel/sdk/resource"
+)
+
+var (
+	agentMetricsPrefixFormat = "agent.googleapis.com/%s"
+	formatter                = func(d *sdkapi.Descriptor) string {
+		return fmt.Sprintf(agentMetricsPrefixFormat, d.Name())
+	}
+)
+
+type enabledReceivers struct {
+	metric map[string]int
+	log    map[string]int
+}
+
+func GetEnabledReceivers(uc *confgenerator.UnifiedConfig) (enabledReceivers, error) {
+	eR := enabledReceivers{
+		metric: make(map[string]int),
+		log:    make(map[string]int),
+	}
+
+	// Logging Pipelines
+	for _, p := range uc.Logging.Service.Pipelines {
+		for _, rID := range p.ReceiverIDs {
+			rType := uc.Logging.Receivers[rID].Type()
+			eR.log[rType] += 1
+		}
+	}
+
+	// Metrics Pipelines
+	for _, p := range uc.Metrics.Service.Pipelines {
+		for _, rID := range p.ReceiverIDs {
+			rType := uc.Metrics.Receivers[rID].Type()
+			eR.metric[rType] += 1
+		}
+	}
+
+	return eR, nil
+}
+
+func InstrumentEnabledReceiversMetric(uc *confgenerator.UnifiedConfig, meter meter.Meter) error {
+	eR, err := GetEnabledReceivers(uc)
+	if err != nil {
+		return err
+	}
+
+	// Collect GAUGE metric
+	gaugeObserver, err := meter.AsyncInt64().Gauge("agent/ops_agent/enabled_receivers")
+	if err != nil {
+		return fmt.Errorf("failed to initialize instrument: %v", err)
+	}
+	_ = meter.RegisterCallback([]instrument.Asynchronous{gaugeObserver}, func(ctx context.Context) {
```

Comment:
I wish I could figure out whether we're supposed to handle errors returned by `RegisterCallback`, but the returned error is not documented anywhere that I can find...

Reply:
Yes, I'm not sure either about handling the error. @dashpole Is there something to do about the error returned by `RegisterCallback` ?

Reply:
Looks like it doesn't ever return an error today, but might in the future: https://github.com/open-telemetry/opentelemetry-go/blob/f535b1e65fe32c3ee15bf27ca8130875122c52f9/sdk/metric/meter.go#L119

Reply:
I would just return the error.

Reply:
Good! Thanks! I updated it to return the error.